File: src/my_py_pkg/my_py_pkg/my_first_node.py
#!/usr/bin/env python3

import rclpy
from rclpy.node import Node

class MyNode(Node):
    def __init__(self):
        super().__init__("py_test")
        self.counter_ = 0
        self.get_logger().info("OOP world")
        self.create_timer(0.5, self.timer_callback)

    def timer_callback(self):
        self.counter_ += 2
        self.get_logger().info("In timer callback, counter = " + str(self.counter_))



def main(args=None):
    rclpy.init(args=args)
    node = MyNode()
    node.get_logger().info("Hello ROS 2 from Python")
    rclpy.spin(node)
    rclpy.shutdown()

if __name__ == "__main__":
    main()

File: src/my_cpp_pkg/src/my_first_node.cpp
#include <rclcpp/rclcpp.hpp>

class MyCppNode:public rclcpp::Node
{
    public:
        MyCppNode():Node("cpp_test"), counter_(0)
        {
            RCLCPP_INFO(this->get_logger(), "C++ OOP Constructor");

            timer_ = this->create_wall_timer(std::chrono::milliseconds(500), std::bind(&MyCppNode::timerCallback, this));
        }

    private:
        void timerCallback()
        {
            RCLCPP_INFO(this->get_logger(), "Inside callback timer, counter = %i", counter_);
            counter_++;
        }

        rclcpp::TimerBase::SharedPtr timer_;
        int counter_;
};


int main(int argc, char ** argv)
{
    rclcpp::init(argc, argv);

    auto node = std::make_shared<MyCppNode>();

    RCLCPP_INFO(node->get_logger(), "Hello from CPP node");
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}